File: src/com/chinmay28/TwoSum.java
package com.chinmay28;

import java.util.HashMap;
import java.util.Map;


public class TwoSum {

    public static void main(String args[]) {
        System.out.println("Hello!");
        Map<Integer, Integer> indexMap = new HashMap();
        indexMap.put(1,2);
        indexMap.put(3,1);

        for (int key: indexMap.keySet()) {
            System.out.println("Key:" + key + " Value:" + indexMap.get(key));
        }

        new Solution().twoSum(new int[]{3,2,4}, 6);
    }
}


class Solution {

    public int[] twoSum(int[] nums, int target) {

        HashMap<Integer, Integer> indexMap = new HashMap<>(nums.length);

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (indexMap.containsKey(complement)) {
                return new int[]{i, indexMap.get(complement)};
            }
            indexMap.put(nums[i], i);
        }
        return null;
    }
}

File: cpp/problems/two_sum.cpp
/*
https://leetcode.com/problems/two-sum/

1. Two Sum

Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

Example:

Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].

*/
#include<iostream>
#include<map>
#include<vector>

using namespace std;


class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        
        map<int, int> indexMap;
        vector<int> result;
        
        for (int i=0; i < nums.size(); i++) {

            int complement = target - nums[i];            
            if (indexMap.end() != indexMap.find(complement)) {
                result.push_back(indexMap[complement]);
                result.push_back(i);
                break;
            } else {
                indexMap[nums[i]] = i;
            }
        }
        return result;
    }
};


int main() {
    vector<int> numbers = {2, 7, 11, 15}; 
    vector<int> result = Solution().twoSum(numbers, 9);

    cout << result[0] << ", " << result[1] << endl;
    return 0;
}